### ISSUE TITLE
Compile Errors

### ISSUE THREAD

**Issue by manjus3** [commschamp/cc.demo1.commsdsl#1]
Hi,

I just came across the comms library and I wanted to  play with it a bit. However when I tried to compile the  demo program (after successfully compiling and installing commsdsl and champion ), I get the following errors:

In file included from /home/manju/development/3rdparty/boost_1_69_0/boost/container_hash/hash_fwd.hpp:13:0,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/type_index/type_index_facade.hpp:13,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/type_index/stl_type_index.hpp:22,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/type_index.hpp:29,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/any.hpp:20,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/program_options/value_semantic.hpp:12,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/program_options/options_description.hpp:13,
                 from /home/manju/development/3rdparty/boost_1_69_0/boost/program_options.hpp:15,
                 from /home/manju/development/sandbox/c++/comms/demo/cc.demo1.commsdsl/examples/server/ProgramOptions.h:4,
                 from /home/manju/development/sandbox/c++/comms/demo/cc.demo1.commsdsl/examples/server/main.cpp:5:
/home/manju/development/3rdparty/boost_1_69_0/boost/type_traits/detail/config.hpp:85:69: error: "__clang_major___WORKAROUND_GUARD" is not defined, evaluates to 0 [-Werror=undef]
       && !BOOST_WORKAROUND(BOOST_MSVC, < 1900) && !BOOST_WORKAROUND(__clang_major__, <= 4)
                                                                     ^
/home/manju/development/3rdparty/boost_1_69_0/boost/config/workaround.hpp:246:10: note: in definition of macro ‘BOOST_WORKAROUND’
        ((symbol ## _WORKAROUND_GUARD + 0 == 0) &&     \
          ^~~~~~
/home/manju/development/3rdparty/boost_1_69_0/boost/type_traits/detail/config.hpp:85:69: error: "__clang_major__" is not defined, evaluates to 0 [-Werror=undef]
       && !BOOST_WORKAROUND(BOOST_MSVC, < 1900) && !BOOST_WORKAROUND(__clang_major__, <= 4)
                                                                     ^
/home/manju/development/3rdparty/boost_1_69_0/boost/config/workaround.hpp:247:9: note: in definition of macro ‘BOOST_WORKAROUND’
        (symbol != 0) && (1 % (( (symbol test) ) + 1)))
         ^~~~~~
^Cexamples/server/CMakeFiles/cc.demo1.commsdsl_server_example.dir/build.make:62: recipe for target 'examples/server/CMakeFiles/cc.demo1.commsdsl_server_example.dir/main.cpp.o' failed
make[2]: *** [examples/server/CMakeFiles/cc.demo1.commsdsl_server_example.dir/main.cpp.o] Interrupt
CMakeFiles/Makefile2:175: recipe for target 'examples/server/CMakeFiles/cc.demo1.commsdsl_server_example.dir/all' failed
make[1]: *** [examples/server/CMakeFiles/cc.demo1.commsdsl_server_example.dir/all] Interrupt
Makefile:129: recipe for target 'all' failed

Thanks
Manju

**Comment by arobenko** [comment]
Hi Manju,
I don't know for sure, but it looks like some boost configuration error. It usually means some kind of mismatch between your version of boost and the compiler. That has little to do with my code. Try to
Google the problem, or try older version of boost. My build works fine with gcc / clang compilers with boost v1.58 on Linux and with v1.65 on Windows.

**Comment by manjus3** [comment]
OK, thanks. I'll try again with Boost 1.58. Currently, I have the latest Boost v1.69, and the latest gcc/clang on Ubuntu 18.04. Googling for the above errors didn't come up with anything, unfortunately.

Thanks,
Manju

**Comment by manjus3** [comment]
Hi Alex,

This is so bizarre. I tried several versions of boost, and I got errors (related to boost) in all of them. Then I tried this command line, bypassing the cmake system:
$g++ server.cpp ProgramOptions.cpp Server.cpp Session.cpp -lboost_program_options -lboost_filesystem -lboost_system -pthread -o main -I ../../../cc.demo1.commsdsl/build/output.tmp/include/

This compiled fine. So, maybe there's something in the CMakeLists.txt that triggers all these boost errors. You might also want to try and compile this particular project using the latest boost and g++ or clang, and see if the CMakeListst.txt needs to be reworked.

**Comment by arobenko** [comment]
I'd also recommend using boost that comes with your Ubuntu system, don't
use any external compilation of boost.

You can also set VERBOSE=1 and analize the command line. There may be some
conflict between boost and your system headers.

On linux system you should be able to run cmake without any extra
parameters and it should be able to find all the necessary libraries

**Comment by manjus3** [comment]
Thanks, Alex. I'll look into that.

**Comment by manjus3** [comment]
OK, so I pulled the default boost libraries using apt on Ubuntu and ran cmake against it, and now everything compiles fine.

On another note, I was wondering if your DSL supports specification of finite-state machines (FSMs) for wire protocols. If not, I'm thinking of extending COMMS to support FSMs, once I'm more familiar with the code base. Many wire protocols are state-based, and so a specification of message formats and states go hand in hand. I thinking of using  Boost.MSM or SML for this. If you have any suggestions on this topic, please let me know. 

Thanks,
Manju

**Comment by arobenko** [comment]
Hi Manju,
The COMMS library is there to facilitate implementation of binary communication protocol. It gives a lot of flexibility on how the message exchange functionality is going to be integrated with business logic. If you want to use state machine go for it. Just use whatever library you like, it should be relatively easy to integrate generated code. 

Please read "**How to Use Defined Custom Protocol**" page from the COMMS library documentation, and "Message Handling" section in particular. I'm not really familiar with state machines from Boost, but I suppose you will need to:

- support polymorphic dispatch in message interface.
- define a variant **comms::GenericHandler** to be a base class for all the classes representing states (use multiple inheritance if needed)
- override appropriate virtual **handle()** member functions for every state class for the messages you want to handle in a particular state.
- implement your own logic of changing states upon reception of some message(s).

Your best info resource is "**How to Use Defined Custom Protocol**" page from the COMMS library documentation. Please read it thoroughly. If something is not clear, please send me a direct e-mail with your questions.

**Comment by manjus3** [comment]
Hi Alex, thanks for pointing me to the document. I'll take a look and see what I need to do for this task. Ideally, I'd like to make the FSM as part of the COMMS DSL language, (which would later get generated as C++ FSM structures, and processing logic), instead of inserting the state machine by hand into the generated code.

**Comment by arobenko** [comment]
Usually it's the opposite, the generated code has limited flexibility and you integrate it into your business logic (and not vice versa). How you implement your business logic is completely up to you. State machines are part of your business logic, not the protocol itself. I for example don't like state machines. They look simple on paper when you design them, but with time the number of states grow (especially when protocol grows) and after a while it becomes a mess (unless the protocol itself is very simple) which is difficult to follow (but this is just my perception, not claiming it to be true for everyone else). Quite often it is not clear whether to introduce extra "if" statements the check various conditions inside the state or create a separate state.

The bottom line, I insist on communication protocol definition and handling state machine(s) being completely independent and I don't want to enforce usage of a particular state machine implementation / generation. If there is any other DSL that can be used to define state machine, use it.
